python_modules/dagster/dagster/_core/instance/__init__.py

```diff
@@ -1332,33 +1352,70 @@ def _log_asset_planned_events(
                                 f" {PARTITION_NAME_TAG}"
                             )
 
+                        partitions_subset = None
                         if partition_range_start or partition_range_end:
                             if not partition_range_start or not partition_range_end:
                                 raise DagsterInvariantViolationError(
                                     f"Cannot have {ASSET_PARTITION_RANGE_START_TAG} or"
                                     f" {ASSET_PARTITION_RANGE_END_TAG} set without the other"
                                 )
 
-                            # TODO: resolve which partitions are in the range, and emit an event for each
+                            if external_partitions_def_data is None:
+                                raise DagsterInvariantViolationError(
+                                    "Must provide external_partitions_def_data to create_run when creating "
+                                    "a run with a partition range."
+                                )
 
-                        partition = (
-                            partition_tag
-                            if check.not_none(output.properties).is_asset_partitioned
-                            else None
-                        )
+                            # For now, yielding materialization planned events for single run backfills
+                            # is only supported on cloud
+                            if (
+                                self.is_cloud_instance
+                                and check.not_none(output.properties).is_asset_partitioned
+                            ):
+                                partitions_def = (
+                                    external_partitions_def_data.get_partitions_definition()
```

Comment:
This might not be super cheap, right? For static partitions it's O(# partitions), and for time window partitions it involves some pendulum stuff. Thoughts on caching the `PartitionsDefinition` in `pipeline_and_execution_plan_cache` instead of the `ExternalPartitionsDefinitionData`?

Another benefit of this would be spreading `ExternalPartitionsDefinitionData` to fewer places – longer term it could be nice to get rid of that and just use `@whitelist_for_serdes` on `PartitionsDefinition` instead.

Reply:
> Thoughts on caching the PartitionsDefinition in pipeline_and_execution_plan_cache instead of the ExternalPartitionsDefinitionData?

You mean passing around the `PartitionsDefinition` instead of the external partitions definition everywhere, right? I'm good with this, this probably saves us some partitions def roundtrips in the asset backfill case

Reply:
> You mean passing around the PartitionsDefinition instead of the external partitions definition everywhere, right?

Exactly

Reply:
Wondering if we need to worry about it being unclear that backcompat dynamic partitions defs won't be passed into `create_run`...

Could name the new param `assets_partitions_def`? What do you think?

Reply:
> Could name the new param assets_partitions_def? What do you think?

That's a good idea

Reply:
ok, Ive gone ahead with the rename

though I changed it to `asset_job_partitions_def` since I think that is a better description